static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+	const data = eventToStandardMessage(event);
+
+	if (data === null) {
+		return;
+	}
+
+	const listener = LISTENERS[data.type];
+
+	if (Array.isArray(listener) && listener.length) {
+		// Because any listener can have side-effects (by unregistering itself),
+		// we run the promise chain on a copy of the `LISTENERS` array.
+		// Hat tip @piuccio
+		const promise =
+			// We offer, but don't impose, the possibility that a listener returns
+			// a value that must be sent back to the calling frame. To do this,
+			// we pass the cumulated returned value as a second argument to each
+			// listener. Notice we don't try some clever way to compose the result
+			// value ourselves, this would only make the solution more complex.
+			// That means a listener can ignore the cumulated return value and
+			// return something else entirely—life is unfair.
+			// We don't know what each callack will be made of, we don't want to.
+			// And so we wrap each call in a promise chain, in case one drops the
+			// occasional fastdom bomb in the middle.
+			listener.reduce<Promise<unknown>>(
+				(func, listener) =>
+					func.then((ret) => {
+						const thisRet = listener(
+							data.value,
+							ret,
+							getIframe(data),
+						);
+						return thisRet === undefined ? ret : thisRet;
+					}),
+				Promise.resolve(),
+			);
+
+		return promise
+			.then((response) => {
+				respond(data.id, event.source, null, response);
+			})
+			.catch((ex) => {
+				reportError(ex, {
+					feature: 'native-ads',
+				});
+				respond(data.id, event.source, formatError(error500, ex), null);
+			});
+	} else if (typeof listener === 'function') {
+		// We found a persistent listener, to which we just delegate
+		// responsibility to write something. Anything. Really.
+		listener(data.value, respond, getIframe(data));
+	} else {
+		// If there is no routine attached to this event type, we just answer
+		// with an error code
+		respond(data.id, event.source, formatError(error405, data.type), null);
+	}
+};
+
+const on = (window: WindowProxy) => {
+	window.addEventListener('message', (event) => void onMessage(event));
+};
+
+const off = (window: WindowProxy) => {
+	window.removeEventListener('message', (event) => void onMessage(event));
+};
+
+/**
+ * Register a callback for a given type of iframe message
+ *
+ * @param type The `type` of message to register against
+ * @param callback The callback to register that will receive messages of the given type
+ * @param options Options for the target window and whether the callback is persistent
+ */
+export const register: RegisterListener = (type, callback, options): void => {
+	if (REGISTERED_LISTENERS === 0) {
+		on(options?.window ?? window);
+	}
+
+	// Persistent LISTENERS are exclusive
+	if (options?.persist) {
+		LISTENERS[type] = callback;
+		REGISTERED_LISTENERS += 1;
+	} else {
+		const listeners = LISTENERS[type] ?? [];
+
+		if (Array.isArray(listeners) && !listeners.includes(callback)) {
```

Comment:
Does this mean that if we register a persistent callback, then attempt to register a non-persistent one, that second registration will have no effect and caller won't be aware of this? And is this the expected behaviour?
If not, this might be another reason to use separate properties for persistent and non-persistent callbacks.

Reply:
> Does this mean that if we register a persistent callback, then attempt to register a non-persistent one, that second registration will have no effect and caller won't be aware of this? 

Yes. I think the behaviour isn't clear at all.

> And is this the expected behaviour?

Good question. I want to leave the implementation close to the original, but this definitely fits in with @mxdvl's suggestions above to ditch persist callbacks entirely, or at least label them.

From my testing it seems like we don't ever actually try to register a persistent then non-persistent callback of the same type, so I think this is safe for now, but definitely needs to be refactored!